File: baekjoon/Divide and Conquer/2263.py
def preOrder(inStart, inEnd, postStart, postEnd):
    if (inStart > inEnd) or (postStart > postEnd):
        return
    root = postOrder[postEnd]
    print(root, end = " ")

    left = position[root] - inStart
    right = inEnd - position[root]

    preOrder(inStart, inStart + left - 1, postStart, postStart + left -1)
    preOrder(inEnd - right + 1, inEnd, postEnd - right, postEnd - 1)

n = int(input())
inOrder = list(map(int, input().split()))
postOrder = list(map(int, input().split()))

position = [0] * (n + 1)
for i in range(n):
    position[inOrder[i]] = i

preOrder(0, n - 1, 0, n - 1)

File: baekjoon/Divide and Conquer/2263.cpp
#include <iostream>
using namespace std;

/*
 * InOrder: Left>Root>Right
 * PreOrder: Root>Left>Right 
 * PostOrder: Left>Right>Root
*/

int inOrder[100001];
int postOrder[100001];
int idxArr[100001];

void preOrder(int inStart, int inEnd, int postStart, int postEnd)
{
	if (inStart > inEnd || postStart > postEnd)
		return;

	int root = postOrder[postEnd];
	int rootIdx = idxArr[root];
	int leftSize = rootIdx - inStart;
	int rightSize = inEnd - rootIdx;
	
	cout << inOrder[rootIdx] << ' ';

	preOrder(inStart, rootIdx - 1, postStart, postStart + leftSize - 1);
	preOrder(rootIdx + 1, inEnd, postStart + leftSize, postEnd - 1);
}

int main(void)
{
	int n;
	cin >> n;

	for (int i = 1; i <= n; i++) {
		cin >> inOrder[i];
		idxArr[inOrder[i]] = i;
	}

	for (int i = 1; i <= n; i++)
		cin >> postOrder[i];

	preOrder(1, n, 1, n);
	cout << "\n";

	return 0;
}
